### PR TITLE
Add CAST (double as decimal)

### PR DESCRIPTION
**Pull request by rui-mo**
A replacement for https://github.com/facebookincubator/velox/pull/5767 from @wuxueyang96.
https://github.com/facebookincubator/velox/issues/8412

### CODE REVIEW

File: velox/common/base/tests/GTestUtils.h [hunk 1]
```diff
@@ -54,6 +54,13 @@
   VELOX_ASSERT_THROW_IMPL(                                     \
       facebook::velox::VeloxRuntimeError, _expression, _errorMessage)
 
+#define VELOX_ASSERT_ERROR_STATUS(_expression, _statusCode, _errorMessage) \
```
Comment (rui-mo) [inline velox/common/base/tests/GTestUtils.h:57]: Add VELOX_ASSERT_ERROR_STATUS if it is needed.

File: velox/expression/tests/CastExprTest.cpp [hunk 2]
```diff
@@ -1992,6 +1992,106 @@ TEST_F(CastExprTest, castInTry) {
...
+      "Cannot cast DOUBLE '1.7976931348623157E308' to DECIMAL(38, 2). Result overflows.");
+  testThrow<double>(
+      DOUBLE(),
+      DECIMAL(38, 2),
+      {std::numeric_limits<double>::lowest()},
+      "Cannot cast DOUBLE '-1.7976931348623157E308' to DECIMAL(38, 2). Result overflows.");
+  testComplexCast(
+      "c0",
+      makeConstant<double>(std::numeric_limits<double>::min(), 1),
+      makeConstant<int128_t>(0, 1, DECIMAL(38, 2)));
```
Comment (rui-mo) [inline velox/expression/tests/CastExprTest.cpp:2081]: Tests for both lowest and min are added.

**Review by rui-mo — COMMENTED** [review]
@mbasmanova Could you help review this PR? Some unmerged changes from https://github.com/facebookincubator/velox/pull/8478 are added in this PR if still needed. Thanks.

File: velox/common/base/tests/GTestUtils.h (hunk 1 above)
Comment (mbasmanova) [reply velox/common/base/tests/GTestUtils.h:57]: CC: @pedroerp

File: velox/common/base/tests/GTestUtils.h [hunk 3]
```diff
@@ -54,6 +54,13 @@
...
+  const auto status = (_expression);                                       \
+  ASSERT_TRUE(status.code() == _statusCode);                               \
```
Comment (mbasmanova) [inline velox/common/base/tests/GTestUtils.h:59]: Would it be possible to include mismatches codes in the error message? What does the failure look like now?

File: velox/expression/CastExpr-inl.h [hunk 4]
```diff
@@ -453,6 +453,42 @@ void CastExpr::applyIntToDecimalCastKernel(
...
 }
 
+template <typename TOutput>
+void CastExpr::applyDoubleToDecimalCastKernel(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& result) {
+  const auto sourceVector = input.as<SimpleVector<double>>();
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:463]: nit: doubleInput

File: velox/expression/CastExpr-inl.h [hunk 5]
```diff
@@ -453,6 +453,42 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+  auto rawResults =
+      result->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto toPrecisionScale = getDecimalPrecisionScale(*toType);
+
+  applyToSelectedNoThrowLocal(context, rows, result, [&](vector_size_t row) {
+    if (sourceVector->isNullAt(row)) {
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:469]: Will this ever be the case? Are we not handling nulls elsewhere?

File: velox/expression/tests/CastExprTest.cpp [hunk 6]
```diff
@@ -1992,6 +1992,106 @@ TEST_F(CastExprTest, castInTry) {
...
+      makeFlatVector<double>({-3333.03, -2222.02, -1.0, 0.00, 100, 99999.99});
+  testComplexCast(
+      "c0",
+      input,
+      makeFlatVector<int64_t>(
+          {-33'330'300, -22'220'200, -10'000, 0, 1'000'000, 999'999'900},
+          DECIMAL(10, 4)));
+
+  // Double to long decimal.
+  testComplexCast(
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:2007]: Can you remind me what does 'complex' mean here? This is not about complex types, is it?

**Review by mbasmanova — APPROVED** [review]
@rui-mo Looks great % some nits.

File: velox/expression/tests/CastExprTest.cpp (hunk 6 above)
Comment (rui-mo) [reply velox/expression/tests/CastExprTest.cpp:2007]: It is similar with testCast with some tiny differences. This is one PR https://github.com/facebookincubator/velox/pull/8254 to combine them inspired by https://github.com/facebookincubator/velox/pull/7377#discussion_r1440093802. cc @kagamiori

Comment (mbasmanova) [reply velox/expression/tests/CastExprTest.cpp:2007]: Got it. Thank you for clarifying. Let's review and land that other PR to reduce confusion.

File: velox/common/base/tests/GTestUtils.h (hunk 3 above)
Comment (rui-mo) [reply velox/common/base/tests/GTestUtils.h]: Added code in the error message. Now the message looks like below.
Code mismatch:
```
Expected error code to be 'Type error', but received 'User error'.
```
Message mismatch:
```
Value of: status.message().find(expectedErrorMessage) != std::string::npos
  Actual: false
Expected: true
Expected error message to contain 'Unknown error.', but received 'Result overflows.'.
```

File: velox/expression/CastExpr-inl.h (hunk 5 above)
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: I checked and found this branch would not be entered. Removed this branch and I will open a PR to remove similar checks of other cast kernels.

File: velox/common/base/tests/GTestUtils.h (hunk 3 above)
Comment (mbasmanova) [reply velox/common/base/tests/GTestUtils.h]: Nice. Thank you.

File: velox/expression/CastExpr-inl.h (hunk 5 above)
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h]: Perfect. Thanks.

**Review by Yuhta — APPROVED** [review]

File: velox/expression/CastExpr-inl.h (hunk 5 above)
Comment (rui-mo) [reply velox/expression/CastExpr-inl.h]: Opened https://github.com/facebookincubator/velox/pull/8557 for this minor fix. Thanks.

**Pull request closed without merging**